DEBUG: Drop custom syslog identifier from journald

sigv opened the pull request:
`SYSLOG_IDENTIFIER` field is removed from logging output to journald. The default behavior will use the program name automatically.

This ensures that if there is rsyslog in place, producing BSD-format syslog output (RFC 3164), then there are no unexpected `[` characters. The resulting syslog output will also be aligned with the behavior of `--with-syslog=syslog` which uses the program name by default.

`SSSD_PRG_NAME` field has been added to the journald log as well, to be consistent with the `DEBUG()` output. This field holds the value that `SYSLOG_IDENTIFIER` had previously and can be used for filtering journal as a drop-in replacement.

---

Initial PR commit replaced square brackets `[]` in `SYSLOG_IDENTIFIER` with parentheses `()`.

Replacing lines forwarded to syslog using rsyslog from

- `sssd[sudo][1234]: `
- `sssd[be[EXAMPLE.COM]][1235]: `

to

- `sssd(sudo)[1234]: `
- `sssd(be(EXAMPLE.COM))[1235]: `

This was dropped after comments from @alexey-tikhonov below, and replaced with the `SYSLOG_IDENTIFIER` to `SSSD_PRG_NAME` change.

mathewhodson commented on src/p11_child/p11_child_common.c:
```diff
@@ -308,7 +308,7 @@ int main(int argc, const char *argv[])
     DEBUG_INIT(debug_level);
 
-    debug_prg_name = talloc_asprintf(NULL, "p11_child[%d]", getpid());
+    debug_prg_name = talloc_asprintf(NULL, "p11_child(%d)", getpid());
```
Shouldn't this continue to use square brackets because it's including a process ID?

sigv replied:
As I see it, Journald uses a combination of `SYSLOG_IDENTIFIER` and `SYSLOG_PID` when producing the final syslog output. They get merged as string `${SYSLOG_IDENTIFIER}[${SYSLOG_PID}]: `.

So if `debug_prg_name` is set to `p11_child[1234]` then `SYSLOG_IDENTIFIER` would be set to `sssd(p11_child[1234])`, resulting in syslog output of `sssd(p11_child[1234])[1234]: `.

Potentially with a greedy regexp matching `\[(.+)\]` you get `1234])1234` as the PID.

That is the reason why I am swapping all the `debug_prg_name` square brackets to parentheses.

mathewhodson commented:
I think the two lines from commit 18233532b72e62452eac6886652fa633ba055d8c also need to be changed.

sigv commented:
> I think the two lines from commit 1823353 also need to be changed.

`SYSLOG_IDENTIFIER=sssd[%s]` in src/util/sss_log.c has been replaced with `SYSLOG_IDENTIFIER=sssd(%s)`. As of now that is the last file in the diff if sorted alphabetically.

`SSSD_PRG_NAME=sssd[%s]` in src/util/debug.c may be retained with square brackets, as it does not relate to [syslog fields](https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html#SYSLOG_FACILITY=). Would you prefer me to swap `SSSD_PRG_NAME` to be `sssd(%s)` as well in this PR?

alexey-tikhonov commented:
Hi @sigv ,

at this point I'm a little bit confused what issue exactly we are trying to solve.

Initially #5431 was talking about the case:
 - SSSD is built with --with-syslog=syslog
 - syslog was trying to parse [] part of `debug_prg_name` as a PID
 - the result was seen in journalctl output as malformed SYSLOG_IDENTIFIER and SYSLOG_PID (and this was the issue)

Is this ^^ correct so far?

This specific case should be resolved with 18233532b72e62452eac6886652fa633ba055d8c as this patch removes ident==debug_prg_name in case "--with-syslog=syslog" and default is used instead. So instead of sssd[sudo] you should see sssd_sudo (no [])


Now this PR seems to talk about different case:
 - SSSD is built with --with-syslog=journald 
 - and then logs are forwarded to syslog (journald composes TAG as SYSLOG_IDENTIFIER+[PID] in the process)
 - then, I guess, syslog gets wrong part of the TAG as a PID?
What is practical issue / consequences?


The thing is, RFC you mention:

1) says "Any non-alphanumeric character will terminate the TAG field and will be assumed to be the starting character of the CONTENT field".
So strictly speaking replacement of [] with () doesn't look as a proper solution.
But I'm somewhat confused here: `man openlog`: "If ident is NULL, the program name is used." and '_' is a perfectly fine in a program name, but it's not alphanumeric character...

2) rfc3164 is obsoleted by rfc5424, and this rfc says "The TAG is now part of the header, but not as a single field.  The TAG has been split into APP-NAME, PROCID, and MSGID." and "HEADER = PRI VERSION SP TIMESTAMP SP HOSTNAME SP APP-NAME SP PROCID SP MSGID", where:
```
   The APP-NAME field SHOULD identify the device or application that
   originated the message.  It is a string without further semantics.
   It is intended for filtering messages on a relay or collector.
```
Take a note SP is used as a separator between header fields.
I don't see how "[]" in APP-NAME violates rfc5424 (please correct me if I'm wrong)

Those two RFC are quite different that's why I ask for more details about your setup.

alexey-tikhonov commented:
Frankly, I'm tempted to just use sd_journal_send() with `"SYSLOG_IDENTIFIER=sssd[%s]", program_invocation_short_name`.

This way we will not change format of logs in files and will use a process name as an SYSLOG_IDENTIFIER, and I think any reasonable syslog implementation should be able to cope with it.

@sumit-bose , @sigv , what do you think about this?

alexey-tikhonov commented:
Btw, we only specify `SYSLOG_IDENTIFIER` in explicit `sss_log()` calls.
In the case jorunald is used as a general logger, we specify `SSSD_PRG_NAME`.
I don't know why is this difference.

sigv commented:
> Initially #5431 was talking about the case:
> - SSSD is built with --with-syslog=syslog
> - syslog was trying to parse [] part of debug_prg_name as a PID
> - the result was seen in journalctl output as malformed SYSLOG_IDENTIFIER and SYSLOG_PID (and this was the issue)
>
> This specific case should be resolved with 1823353 as this patch removes ident==debug_prg_name in case "--with-syslog=syslog" and default is used instead. So instead of sssd[sudo] you should see sssd_sudo (no [])

All correct on the summary. The reason for the improper `SYSLOG_IDENTIFIER` and `SYSLOG_PID` was that systemd-journald was parsing syslog back into journald. It was picking up misformated lines and mis-parsing the `TAG[PID]` combination there.

The mitigation is partially a combination of 00e7b1ada3d1c1071eac79b65c17cd2701c2ae6a and 18233532b72e62452eac6886652fa633ba055d8c with one removing `sssd[` from one place and adding `sssd[` in another. I did close that opened #5431 issue ticket on the tracker here as a result, as it was determined to be a flaw in historical versions.

> then, I guess, syslog gets wrong part of the TAG as a PID?
> What is practical issue / consequences?

This is where me being interested in switching `--with-syslog=syslog` to `--with-syslog=journald` comes into play. As of now, by default in Ubuntu (and I would guess Debian as well), `/var/log/syslog` gets forwarded to from journald, with rsyslog. As you can see from various sources, rsyslog utilizes BSD syslog format as in RFC 3164 instead of the RFC 5424 format.

However for Ubuntu Core Development Team brought up the fact that on Ubuntu `/var/log/syslog` output as `sssd[sudo][1234]:` after the current proposed changes. If someone is working with the syslog output then they will have a bad time when trying to parse PID, which with a greedy (default) regexp for `/\[.*\]/` or `.split('[')[0].split(']')[0]` results with `"[sudo][1234]"`. 

> So strictly speaking replacement of [] with () doesn't look as a proper solution.

The informational RFC 3164 from year 2001 was not a perfectly formal standard and there are adaptations to it. Parsers generally care about the `:` and `[` characters as separators in the start of "content". The goal here is to avoid `sudo` bit leaking into `PID`.

> I don't see how "[]" in APP-NAME violates rfc5424 (please correct me if I'm wrong)

In RFC 5424 both `sssd[sudo]` and `sssd(sudo)` could be considered valid indeed. However, RFC 3164 is still in use in the wild. And usual parsers there handle `[]` in a particular way, as PID separators.

> Frankly, I'm tempted to just use sd_journal_send() with "SYSLOG_IDENTIFIER=sssd[%s]", program_invocation_short_name.

`sssd[sssd_sudo]` still has a `[` character. That is the concern. If you want, then it could be replaced with `SYSLOG_IDENTIFIER=sssd(%s)` and `program_invocation_short_name` instead of `debug_prg_name`.

But at that point, as you have `sssd(sssd_sudo)` you might as well just use `program_invocation_short_name` (`sssd_sudo`) directly.

Not sure if this proposal is improving the sitation as a result.

> This way we will not change format of logs in files and will use a process name as an SYSLOG_IDENTIFIER, and I think any reasonable syslog implementation should be able to cope with it.

There is a change in any way. The change proposed by this PR just swaps `[` characters with `(` characters. If someone was parsing them previously, they can keep the same logic they did by swapping the two.

"any reasonable syslog implementation" does not deal with the current code due to the (old) RFC 3164 expecting `[` to separate PIDs, and causes pain. Does `be` report the domain in its program name even? The change to process name causes more change than just swapping `[` to `(`. 

> I don't know why is this difference.

The difference is in how rsyslog picks up the journald message and ships it to BSD-style syslog (as the old RFC 3164).

sigv commented:
To summarize, journald may be forwarded to syslog using rsyslog.
The older BSD-style RFC 3164 is used by it resulting in `SYSLOG_IDENTIFIER[SYSLOG_PID]:`.
For current code, that ends up being `sssd[sudo][1234]:` from which some may parse PID as `sudo][1234`.
This PR therefore proposed `sssd(sudo)[1234]:` as a safer replacement, where parsing PID always results with `1234`.

sigv commented:
And to highlight it further out of the large message, this BSD-formatting issue is a case for Ubuntu, and I would guess Debian as well but haven't checked directly. Any syslog forwarder that follows the newer RFC 5424 format does not get negatively affected and can still function with `(` the same way as with `[`.

sigv commented:
Fedora opted to drop this kind of setup that Ubuntu has back in Fedora 20 (2013), with [Changes/NoDefaultSyslog](https://fedoraproject.org/wiki/Changes/NoDefaultSyslog), recommending manual `rsyslog` installation for syslog-related files.

alexey-tikhonov commented:
> > Frankly, I'm tempted to just use sd_journal_send() with "SYSLOG_IDENTIFIER=sssd[%s]", program_invocation_short_name.
> 
> `sssd[sssd_sudo]` still has a `[` character. That is the concern. If you want, then it could be replaced with `SYSLOG_IDENTIFIER=sssd(%s)` and `program_invocation_short_name` instead of `debug_prg_name`.

Sorry for mistype (bad copy&paste).
I meant `"SYSLOG_IDENTIFIER=%s", program_invocation_short_name` of course.

> But at that point, as you have `sssd(sssd_sudo)` you might as well just use `program_invocation_short_name` (`sssd_sudo`) directly.

Exactly.

> Not sure if this proposal is improving the sitation as a result.

Why? This removes "[]" from journal logs. Doesn't this address the issue?

> > This way we will not change format of logs in files and will use a process name as an SYSLOG_IDENTIFIER, and I think any reasonable syslog implementation should be able to cope with it.
> 
> There is a change in any way. 

No. What I propose doesn't affect logs in files.

> > I don't know why is this difference.
> The difference is in how rsyslog picks up the journald message and ships it to BSD-style syslog (as the old RFC 3164).

This is a consequence. That doesn't answer my question.

SSSD has two kind of loggers: 
 1) DEBUG(...) that is managed via `--logger=files|journal|...` option
 2) sss_log() that always logs to system journal (not files) for most important messages

In case --logger=journal (1) uses `SSSD_PRG_NAME`
(2) uses`SYSLOG_IDENTIFIER`

I guess there is a reason behind this but didn't yet figure this out.


> Does `be` report the domain in its program name even? The change to process name causes more change than just swapping `[` to `(`.

No, it doesn't. But we have `"SSSD_DOMAIN=%s", domain` there. This might be enough (or not, depending on journal being used...)

sigv commented:
> I meant "SYSLOG_IDENTIFIER=%s", program_invocation_short_name of course.

I see. Yes, it does make sense, and aligns with the `--with-syslog=syslog` behavior.

>>> I don't know why is this difference.
>>
>> The difference is in how rsyslog picks up the journald message and ships it to BSD-style syslog (as the old RFC 3164).
>
> This is a consequence. That doesn't answer my question.

If the question was about the difference between `SYSLOG_IDENTIFIER` and `SSSD_PRG_NAME`, then the former is being changed for the BSD-style syslog output as described previously. The latter does not get re-packaged into the content in the same way.

If there is a further question that should be answered, I think I am missing it.

---

Should I then restructure the PR to not have any changes to `debug_prg_name` and instead have simply one line removed and one added as follows?

```diff
diff --git a/src/util/sss_log.c b/src/util/sss_log.c
index c6b7435c6..123456789 100644
--- a/src/util/sss_log.c
+++ b/src/util/sss_log.c
@@ -107,7 +107,7 @@ static void sss_log_internal(int priority, int facility, const char *format,
                     "SSSD_DOMAIN=%s", domain,
                     "PRIORITY=%i", syslog_priority,
                     "SYSLOG_FACILITY=%i", LOG_FAC(facility),
-                    "SYSLOG_IDENTIFIER=sssd[%s]", debug_prg_name,
+                    "SYSLOG_IDENTIFIER=%s", program_invocation_short_name,
                     NULL);

     free(message);
```

alexey-tikhonov commented:
> Should I then restructure the PR to not have any changes to `debug_prg_name` and instead have simply one line removed and one added as follows?
> +                    "SYSLOG_IDENTIFIER=%s", program_invocation_short_name,

Btw, this is glibc/_GNU_SOURCE specific.

It appears more simple solution is to remove `SYSLOG_IDENTIFIER` completely: systemd-journal uses  program_invocation_short_name automatically in this case (I did a quick test and https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html seems to confirm this: "Note that the tag is usually derived from glibc's  program_invocation_short_name variable")

This will also make implementation of `sss_log_internal()` for WITH_JOURNAL case consistent with SYSLOG case and with `journal_send()` used in case of DEBUG() & logger==journal (and btw imo it makes sense to add `"SSSD_PRG_NAME=sssd[%s]", debug_prg_name` to `sss_log_internal()`, again, to be consistent).

I'd like to hear other maintainers opinion though.

sigv commented:
Updated the ticket to represent the current state of the discussion, proposing renaming of `SYSLOG_IDENTIFIER` to `SSSD_PRG_NAME`.

alexey-tikhonov commented:
Yes, that's what I meant, thank you.

@sumit-bose , @pbrezina , what do you think?

sigv commented:
Following merge, this should also be picked in to the `sssd-1-16` branch, as it affects the `1.x` series in the same way.
`DEBUG()` already has `SSSD_PRG_NAME` on `sssd-1-16`.

sumit-bose commented:
> Yes, that's what I meant, thank you.
> 
> @sumit-bose , @pbrezina , what do you think?

Hi,

I'm fine with this, ACK.

bye,
Sumit

pbrezina commented:
Does this resolve: https://github.com/SSSD/sssd/issues/4590?

alexey-tikhonov commented:
> Does this resolve: #4590?

Seems so. I was not aware of this ticket.

sigv commented:
That does appear to be the case based on a cursory glance indeed, as the concern listed concern in the ticket is being resolved:

> **Comment from [erempel](https://pagure.io/user/erempel) at [2017-11-09 18:38:16](https://pagure.io/SSSD/sssd/issue/3566#478261)**
> Regardless of what I think, the documentation for the journald fields states
> SYSLOG_IDENTIFIER - Syslog compatibility field containing the identifier string (i.e. "tag")
> The tag that sssd is using is not syslog compatible.

Should I force-push a `Resolves:` line in the commit, or will someone handle that during a manual push?

alexey-tikhonov commented:
> 
> Should I force-push a `Resolves:` line in the commit, or will someone handle that during a manual push?

Yes, please add following line to the commit message:

```
Resolves: https://github.com/SSSD/sssd/issues/4590
```

sumit-bose commented:
> > Should I force-push a `Resolves:` line in the commit, or will someone handle that during a manual push?
> 
> Yes, please add following line to the commit message:
> 
> ```
> Resolves: https://github.com/SSSD/sssd/issues/4590
> ```

Hi,

while you are working on the commit message I wonder if you can add a `:relnote:` tag as well mentioning the change in the journald meta data? This will help to not forget this change in the release notes.

bye,
Sumit

sigv commented:
> Yes, please add following line to the commit message:

Done.

> while you are working on the commit message I wonder if you can add a :relnote: tag as well mentioning the change in the journald meta data? This will help to not forget this change in the release notes.

Done, I believe.
@sumit-bose, could you please confirm the format is as expected?

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/5477

* `master`
    * 3c8919e85d5afa99c3b30eba51e2f5e24d88fdcb - DEBUG: Drop custom syslog identifier from journald
* `sssd-1-16`
    * 0e80b808e26f8837a4375b2ce42351de3180c340 - DEBUG: Drop custom syslog identifier from journald

pbrezina commented:
I altered the release notes little bit and pushed the patch. Thank you for your contribution.
